### DISCUSSION TITLE
If I have a long list of citations, how would I find them in CL?

### DISCUSSION THREAD

**Discussion by mlissner** [freelawproject/courtlistener#3293] [Q&A]
I have a spreadsheet of citations that I want to marry to CourtListener IDs. How do I do that?

**Comment by mlissner** [comment]
I think my go-to tool is our citation finder:

https://www.courtlistener.com/c/

The great thing about it is that it is *very* robust against weird citations. No matter if you have a historical citation, one with a typo or some other messy thing, it should work, if we have the citation.

You can, of course, just do this by hand, but that's no fun. To do it programmatically, it has a basic API we hope to document in #2714, but haven't yet gotten to. 

Here's how it works. Say I have the citation for [Roe v. Wade](https://www.courtlistener.com/opinion/108713/roe-v-wade/), `410 U.S. 113`. 

I can do a GET to this endpoint like this: 

```
curl -v 'https://www.courtlistener.com/c/U.S./410/113/'
```

That'll return a 302 that takes you do:

```
↪ curl -v 'https://www.courtlistener.com/c/U.S./410/113/'
> GET /c/U.S./410/113/ HTTP/2
> Host: www.courtlistener.com
< location: /c/us/410/113/
```

You can see that your reporter (the `U.S.` part of it) is now corrected to a lowercased version, OK. So let's try *that* URL:

```
↪ curl -v 'https://www.courtlistener.com/c/us/410/113/'
> GET /c/us/410/113/ HTTP/2
> Host: www.courtlistener.com
< location: /opinion/108713/roe-v-wade/
```

And now you have the URL of the case, which has the cluster ID of `108713`. 

From there you can get the docket ID or the opinion ID. The docket ID you can get with:

```
↪ curl --silent 'https://www.courtlistener.com/api/rest/v3/clusters/108713/' | jq '.docket_id'
"https://www.courtlistener.com/api/rest/v3/dockets/488071/"
```

The opinion ID you can get with:

```
↪ curl --silent 'https://www.courtlistener.com/api/rest/v3/opinions/?cluster=108713&fields=id'
{
  "count": 4,
  "next": null,
  "previous": null,
  "results": [
    {
      "id": 9425159
    },
    {
      "id": 9425158
    },
    {
      "id": 9425157
    },
    {
      "id": 108713
    }
  ]
}
```

I hope that helps. Let me know what other questions come up or remain!

**Reply by mlissner** [reply]
One other thought here. Citations aren't always unique and aren't always known. Therefore, we respond with one of three possible HTTP status codes:

 - 302, Temporary Redirect, as above, indicates we have the case, and points you to it.
 - 300, Multiple Choices, indicates the citation is not unique in our system, and that you need to disambiguate the results somehow.
 - 404, Not Found, means we don't have the citation.

**Reply by murtazart** [reply]
Thanks @mlissner this is really helpful. 

This API requires the user to split up the citation like so: `U.S./410/113/`. Is there another API that can accept an unstructured citation input like `410 U.S. 113` to produce the same result?

**Reply by mlissner** [reply]
You know, I opened a bug after writing this because I was annoyed that wasn't possible: #3295. I didn't do a full investigation because I wasn't sure exactly what it would take to fix, but since it bugged you too, I went and fixed it just now. I'm waiting for tests to pass, and when they do the code will be automatically deployed. You can watch that here: https://github.com/freelawproject/courtlistener/pull/3312. This will take about 30 minutes.

Once that's complete, you can do a POST like this:

```
curl -X POST https://www.courtlistener.com/c/ --data-raw 'reporter=22 U.S. 44'
```

That will parse the citation correctly, fix any errors it has, and redirect you to the GET request endpoint:

```
< Location: /c/22%20U.S.%2044/
```

So once this deploys, I think you'll be in good shape to go.

**Reply by murtazart** [reply]
Thanks @mlissner this was super useful. Another question: Does this API also throttle requests?

Context:
I iterated over a CSV containing 100 citations, and I started getting `502` error at around 60th request

![image](https://github.com/freelawproject/courtlistener/assets/146972594/62e191d1-9dcf-46ce-951e-9db9f366a4f7)

**Reply by mlissner** [reply]
We've been fighting off bots this past week, and this must have gotten caught up in the fight. Sorry for the trouble. I'll loosen the bot control on this url tomorrow. Sorry!

**Reply by murtazart** [reply]
no worries, and thanks a lot!

**Reply by mlissner** [reply]
I'm sorry, we spent the day fighting bots and we weren't able to open this one more widely yet. I'll get this fixed tomorrow, but it may need a new URL.

**Reply by murtazart** [reply]
no problem, I was just about to write that I'm now seeing `403:Forbidden` error on using citation finder, probably because the API is disabled?

**Reply by mlissner** [reply]
Not disabled, but everything that's not behind `/api/rest/*` is now heavily throttled unfortunately. This means it's time to make the citation lookup API better, but it'll take a sec.

**Reply by mlissner** [reply]
OK, I believe the endpoint is now open to bots once again. Apologies for the delay and the trouble. (AWS is powerful, but difficult!)

**Reply by mlissner** [reply]
https://github.com/freelawproject/courtlistener/issues/3337 is where we will work on building a proper API for this, but it will be a while until it reaches the top of the queue.

**Reply by tribblelabs** [reply]
If I want to get from a list of citations to the full text of that citation's opinion, what is the best way for me to do that? Right now I am thinking that I can use the post request you describe here to get the opinion id (although, I am parsing it from the url that I am forwarded to, which seems a little sketchy) and then using that opinion_id I can hit the api? I'm a bit discombobulated, thanks in advance. My starting point is a list of reporter citations and my end goal is full text + meta data (like case name, reporter info, date, etc). Many thanks for all the great work you all do!

**Reply by mlissner** [reply]
Yes, for now this is the best API for that purpose, until we convert it to a proper API. Hopefully we will do that soon. Apologies for the delay, we've got a long backlog at the moment.

**Reply by tribblelabs** [reply]
I'm getting a whole bunch of 202 responses, is that part of the bot fighting? Any tips? Seems to stop when I use VPN, but then quickly resume again after 5-10 requests

**Reply by flooie** [reply]
Are you using your API TOKEN

**Reply by mlissner** [reply]
This endpoint doesn't need a token, Bill. He's just using the /c/ endpoint and sending it POSTs. 202 responses surprise me, because they're kind of an obscure status code and not one that I _think_ we use normally.

Unfortunately, I'm on vacation until the 8th, so I won't be able to look into this, but if I'm honest, the best solution is to complete https://github.com/freelawproject/courtlistener/issues/3337, so that we no longer direct people to this funky approach. That'd fix your issue and everybody's in the future. 

Any chance you want to take a stab at that? I expect it wouldn't take too much work — if you're familiar with Django REST Framework. If you're not, then it's probably not worth learning and perhaps we just need to bump the priority on this endpoint so we get it done for you sooner.

What's your timeline? Do you think you might be interested in building the API you seek?

**Reply by tribblelabs** [reply]
I am interested in doing this, for sure. I used to be very familiar with DRF (worked on a project in that world professionally for two years) but I'll admit its been a few years. That was 2019. I would love to get a little more guidance on what is needed there and I would be happy to take a stab at it. 

Also, unrelated - it appears to me that the search endpoint returns some data that the opinion and cluster endpoints do not. For example, I used to have a script that did a search based on the case name and year. It would narrow down the results (the goal was to search such that only one result matched the query) and then from that result my script would pull the following information: 

            output_dict['courtlistener_id'] = data['results'][0]['id']
            output_dict['courtlistener_citation_arr'] = data['results'][0]['citation']
            output_dict['courtlistener_cited_by'] = data['results'][0]['citeCount']
            output_dict['courtlistener_casename'] = data['results'][0]['caseName']
            output_dict['courtlistener_casename_short'] = data['results'][0]['caseNameShort']
            output_dict['courtlistener_court'] = data['results'][0]['court']
            output_dict['courtlistener_sibling_id_arr'] = data['results'][0]['sibling_ids']
            output_dict['absolute_url'] = data['results'][0]['absolute_url']
            
Now that I have matured to the cluster/[id] endpoint, though, I find myself missing some of these keys. Is that right? Or perhaps I am just overlooking the information. 

Finally, (and I am sorry to lay so much into one post), if I do grab an opinion by the opinion_id or the cluster_id, what is the best way for me to grab the entire text? I'm sure this is moronic, but I am not seeing the text reliably from the api, so I am going to the 'absolute_url' and scraping it old school style from the html with selenium. Is that the right approach? My thought is that it isn't, though I can make it work. Each of my questions betrays more of my idiocy, but alas, here I am. Enjoy your vactation! Give me whatever guidance you can on #3337 and I will see what I can do! If I do write something, it will need to be reviewed thoroughly :) As always - thanks for everything you do!

**Reply by flooie** [reply]
@tribblelabs - 
1. you are correct that the search endpoint does provide different information 
2. You should familiarize yourself with the cluster/opinion relationship.  The opinion endpoint represents individual judicial thoughts (ie. majority/dissent/concurrence etc).  So if you have a cluster - you need to look at the `sub opinions` or `opinion_ids` to obtain the text.  
3. Inside each opinion object there are multiple text fields all or just one of which could contain the opinion content.  it depends on where the original opinion content was sourced from.

**Reply by mlissner** [reply]
Thanks all! A few responses...

> 1. you are correct that the search endpoint does provide different information

@tribblelabs, can you say which fields are in /search that you're looking for? Everything should be _somewhere._

> 3. Inside each opinion object there are multiple text fields all or just one of which could contain the opinion content. it depends on where the original opinion content was sourced from.

You can see more about which one to choose here: 

https://www.courtlistener.com/help/api/rest/#opinion-endpoint

---

I'll comment on the other issue to get that ball rolling. Thanks for volunteering to help with it. It's something we're hearing a lot of demand for.

**Reply by tribblelabs** [reply]
@flooie when i go to an opinion endpoint, should I be grabbing the text via the html? For example at https://www.courtlistener.com/api/rest/v3/opinions/108713/ - I'm sure it varies by case but for that one specifically should I just key into ['html'] and grab that string and parse it? Thanks in advance

**Reply by flooie** [reply]
html_with_citaitons or xml_harvard are probably your best bets.  But it varies on a case by case basis depending on where the data or multiple data sources come from.  

It's also very important to understand the data structure from opinion cluster to opinion if you are trying to build the full opinion back

**Reply by purplefloyd14** [reply]
@flooie thats very helpful thank you. Yes I am wrapping my head around that structure, seems to be very useful. 

I have a question: I noticed that there are some cases that have tabs separating the sub-opinions (I'm talking about the UI). Sometimes the default/main tab is 'combined', most of the time it seems, but sometimes its "lead". 

here is my question: are there cases without tabs, where multiple sub opinions exist? In that scenario, are all sub-opinions (like dissent and concurrences) all present in that main text content (I assume they are), and are they represented in the api as a single sub-opinion or are they subdivided in the api? 

I know some such cases must exist but I am having trouble finding one now to inspect its api cluster endpoint. Thanks
